### PR TITLE
Backport bclk clock control to GLK branch

### PR DESCRIPTION
**Pull request by brentlu**
This PR is for ODM/codec vendor validation and review

The series of patches are originated from #4219 which is working on master branch. Since GLK branch is ancient, the implementation of bclk clock control is much simpler here but the idea is the same.

### CODE REVIEW

**Comment by lgirdwood** [comment]
@slawblauciak @mwasko fyi.

**Comment by plbossart** [comment]
let's wait that the code is agreed on with the main branch, this should be a backport and aligned with main. 

I would also like the topology changes to be applied on the main branch so that we can test all this with CI. @bardliao can you take the topology patches and apply them to your PR #4219 ?

**Comment by brentlu** [comment]
@plbossart yes, we will wait main branch finalized and see what's need to be ported to here. I've validated this PR on customer device and this is the result from oscilliscope:

1. both bclk and fsync start toggling in hw_param with correct frequency
2. both bclk and fsync disabled in hw_free.

As a side effect, this PR also fixes da7219 which PLL locks on fsync...

**Comment by plbossart** [comment]
> As a side effect, this PR also fixes da7219 which PLL locks on fsync...

That's why I'd like this change to become the default for all I2S codecs ;-)

**Comment by bardliao** [comment]
> let's wait that the code is agreed on with the main branch, this should be a backport and aligned with main.
> 
> I would also like the topology changes to be applied on the main branch so that we can test all this with CI. @bardliao can you take the topology patches and apply them to your PR #4219 ?

Yes, I took the topology patches with slight modification to PR #4219

**Comment by brentlu** [comment]
update to sync with https://github.com/thesofproject/sof/pull/4288

**Comment by brentlu** [comment]
measure the signal of this backport:

fsync starts with data with one zero sample ahead

tek00000

fsync continues to toggle after bclk is disabled for 5.84us

tek00001

**Comment by plbossart** [comment]
Thanks @brentlu just to be clear.
Are you saying the results are now OK and all problems with CS42L42 are solved with the PR?

File: src/drivers/intel/cavs/ssp.c [hunk 1]
```diff
@@ -126,6 +126,51 @@ static int ssp_context_restore(struct dai *dai)
...
+{
+	struct ssp_pdata *ssp = dai_get_drvdata(dai);
+
+	if (!(ssp->params.clks_control &
+		SOF_DAI_INTEL_SSP_CLKCTRL_BCLK_ES))
+		return 0;
+
+	spin_lock(&dai->lock);
+
+	trace_ssp("ssp_hw_free()");
```
Comment (lyakh) [inline src/drivers/intel/cavs/ssp.c:163]: this doesn't have to be under the lock

Comment (brentlu) [reply src/drivers/intel/cavs/ssp.c:163]: i think we need the lock to protect the ssp->state

**Comment by brentlu** [comment]
@plbossart both ODM and codec vendor could not find problem in this backport. Codec vendor still need to check the performance since bclk frequency is changed but basic function is ok.

**Comment by plbossart** [comment]
> @plbossart both ODM and codec vendor could not find problem in this backport. Codec vendor still need to check the performance since bclk frequency is changed but basic function is ok.

 Good to know, thanks @brentlu 
I'll try to finish up the main branch this week, too many things to do at the moment.

File: src/drivers/intel/cavs/ssp.c (hunk 1 above)
Comment (lyakh) [reply src/drivers/intel/cavs/ssp.c:163]: @brentlu I meant the `trace_ssp("ssp_hw_free()");` line

Comment (brentlu) [reply src/drivers/intel/cavs/ssp.c:161]: it's done. thanks for review

**Comment by brentlu** [comment]
based on the https://github.com/thesofproject/sof/pull/4288 with topology changes to use 2.4MHz bclk and 24-bit sample width

**Comment by lgirdwood** [comment]
@mwasko @slawblauciak this one is for you.

**Comment by mwasko** [comment]
@brentlu I would recommend to create a separate branch for this change to avoid mess on release code that took place quite a long time ago. If you want to apply this fix on top of glk-012-stable-branch then I would like to propose to create glk-012-hot-fix-1 branch for that.

**Comment by brentlu** [comment]
Hi @mwasko,
Sure, no problem.

**Pull request merged**